src/transaction/transaction.rs

```diff
@@ -657,34 +689,89 @@ impl Transaction {
+    async fn check_allow_operation(&self) -> Result<()> {
+        let status = self.status.read().await;
+        match *status {
             TransactionStatus::ReadOnly | TransactionStatus::Active => Ok(()),
             TransactionStatus::Committed
             | TransactionStatus::Rolledback
             | TransactionStatus::StartedCommit
-            | TransactionStatus::StartedRollback => Err(Error::OperationAfterCommitError),
+            | TransactionStatus::StartedRollback
+            | TransactionStatus::Dropped => Err(Error::OperationAfterCommitError),
         }
     }
 
     fn is_pessimistic(&self) -> bool {
         matches!(self.options.kind, TransactionKind::Pessimistic(_))
     }
+
+    async fn start_auto_heartbeat(&mut self) {
+        let c_status = self.status.clone();
+        let primary_key = self
+            .buffer
+            .get_primary_key()
+            .await
+            .expect("Primary key should exist");
+        let start_ts = self.timestamp.clone();
+        let region_backoff = self.options.retry_options.region_backoff.clone();
+        let rpc = self.rpc.clone();
+
+        let heartbeat_task = async move {
+            loop {
+                tokio::time::sleep(DEFAULT_HEARTBEAT_INTERVAL).await;
+                {
+                    let status = c_status.read().await;
+                    if matches!(
+                        *status,
+                        TransactionStatus::Rolledback
+                            | TransactionStatus::Committed
+                            | TransactionStatus::Dropped
+                    ) {
+                        break;
+                    }
+                }
+                let current_ts = rpc.clone().get_timestamp().await?;
+                let request = new_heart_beat_request(
+                    start_ts.clone(),
+                    primary_key.clone(),
+                    (current_ts.physical - start_ts.physical) as u64 + DEFAULT_LOCK_TTL,
+                );
+                let plan = PlanBuilder::new(rpc.clone(), request)
+                    .single_region()
+                    .await?
+                    .retry_region(region_backoff.clone())
+                    .plan();
+                plan.execute().await?;
+            }
+            Ok::<(), Error>(())
+        };
+
+        tokio::spawn(async {
+            if let Err(err) = heartbeat_task.await {
+                error!("Error: While sending heartbeat. {}", err);
+            }
+        });
+    }
 }
 
 impl Drop for Transaction {
     fn drop(&mut self) {
-        if !std::thread::panicking() && self.status == TransactionStatus::Active {
-            match self.options.check_level {
-                CheckLevel::Panic => {
-                    panic!("Dropping an active transaction. Consider commit or rollback it.")
+        {
+            let status = futures::executor::block_on(self.status.read());
+            if !std::thread::panicking() && *status == TransactionStatus::Active {
```

Comment:
I think you should check if the thread is panicking first and early return in that case.